fault injection: stop counting as active fault after delay elapsed

kailun-qin opened the pull request:
The current fault injection filter continues to count the injected delay as an active fault even
after it has elapsed. This makes the output statistics incorrect and impacts the max number of
consecutive faults allowed, particularly for long-lived streams (e.g., a gRPC streaming call).

This patch decreases the active fault count when the delay fault is the only active and has gone
finished.

Fixes https://github.com/envoyproxy/envoy/issues/13069

Signed-off-by: Kailun Qin <kailun.qin@intel.com>

Risk Level: N/A
Testing: Unit tests and regression tests
Docs Changes: N/A
Release Notes: N/A

markdroth commented on test/extensions/filters/http/fault/fault_filter_test.cc:
```diff
@@ -580,7 +580,11 @@ TEST_F(FaultFilterTest, FixedDelayDeprecatedPercentAndNonZeroDuration) {
+
   timer_->invokeCallback();
+
+  EXPECT_EQ(0UL, config_->stats().active_faults_.value());
```
The behavior change here is not just for stats; it's also about the max number of consecutive faults allowed.  I suggest adding a test that sets the max number of faults to 1 and delays 100% of requests, starts a streaming call, and after the streaming call has been delayed, starts a second call.  That second call should also be delayed (without this fix, it would not be).

kailun-qin replied:
Thanks for the comments! I do agree.
One question - should we add this kind of test to sandbox level or were you suggesting it for unit level?

markdroth replied:
I don't have a strong feeling about what kind of test is best here.  I'll let one of the Envoy maintainers answer that.

kailun-qin replied:
@asraa Would you please kindly help give some advice here so that we could progress this PR? Thanks.

alyssawilk replied:
I think a unit test that you can
start request 1 with a fault delay
finish request 1 delay (but not request 1)
start request 2 and have the fault delay kick in
would be fine.  As mark said you'll want to make sure you set the limit on active faults - generally when writing a regression test like this I double check to make sure it's testing the right thing by making sure it fails without my changes :-)

kailun-qin replied:
Thanks for the hints and suggestions! Updated with a new attempt.

alyssawilk replied:
Looks great!  Do you mind adding a regression test that would catch issues like this in future?
I think you can mostly crib off your new test, only have a response limiter active too, and make sure it doesn't free up.

kailun-qin replied:
I'd love to add.
However, it seems to me that making a second request w/o ending the first one would be a little bit hard to do in integration tests (e.g., for new streams, the codec will make sure there's no pending response).
Would you please kindly help point me to some reference if there's any?

kailun-qin replied:
> Looks great! Do you mind adding a regression test that would catch issues like this in future?
> I think you can mostly crib off your new test, only have a response limiter active too, and make sure it doesn't free up.

Let me know if the new test added matches the intention of the suggestion here. Thanks.

kailun-qin commented:
Dear code owners, PTAL, thanks! @rshriram @alyssawilk

alyssawilk commented on source/extensions/filters/http/fault/fault_filter.cc:
```diff
@@ -445,6 +445,13 @@ void FaultFilter::postDelayInjection(const Http::RequestHeaderMap& request_heade
   } else {
+    // Should not continue to count as an active fault after the delay has elapsed
+    // if no other type of fault is active/injected
+    if (response_limiter_ == nullptr && fault_active_) {
```
I don't think this guarantees there's no active fault processing, couldn't there be a response rate limiter still applying?  Maybe have a activeFaults() helper function which checks for delay timer and response_rate_limit_, and only decrements if they're both gone?

kailun-qin replied:
Thanks for the comments. Added with a new helper to check both delay timer and response_rate_limit_.

alyssawilk reviewed (commented):
Thanks for tackling this issue and sorry about the review delay - I'll take over since it looks like @rshriram hasn't picked it up.

alyssawilk commented on test/extensions/filters/http/fault/fault_filter_test.cc:
```diff
@@ -588,6 +588,86 @@ TEST_F(FaultFilterTest, FixedDelayDeprecatedPercentAndNonZeroDuration) {
   EXPECT_EQ(0UL, config_->stats().active_faults_.value());
 }
 
+// Verifies that 2 consecutive delay faults be allowed with the max number of faults set to 1
```
Sorry, should have added a comment before, but can you add this as a bug fix in 
docs/root/version_history/current.rst?

kailun-qin replied:
Sure, will do. Thanks.

alyssawilk reviewed (commented):
Code looks good - just test/docs nits left!

One meta-request, can you not force push?  If you merge rather than rebase, github handles reviews better.

alyssawilk commented:
Oh, I meant in your unit tests, that'd be fine.

On Tue, Feb 2, 2021 at 7:55 AM Kailun Qin <notifications@github.com> wrote:

> *@kailun-qin* commented on this pull request.
> ------------------------------
>
> In test/extensions/filters/http/fault/fault_filter_test.cc
> <https://github.com/envoyproxy/envoy/pull/14762#discussion_r568579215>:
>
> >    timer_->invokeCallback();
> +
> +  EXPECT_EQ(0UL, config_->stats().active_faults_.value());
>
> I'd love to add.
> However, it seems to me that making a second request w/o ending the first
> one would be a little bit hard to do in integration tests (e.g., for new
> streams, the codec will make sure there's no pending response).
> Would you please kindly help point me to some reference if there's any?
> Thanks a lot.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/14762#discussion_r568579215>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AELALPOUNLUZGTOQXD35UX3S47Y4LANCNFSM4WKKWTEQ>
> .
>

kailun-qin commented:
> Code looks good - just test/docs nits left!
> 
> One meta-request, can you not force push? If you merge rather than rebase, github handles reviews better.

Many thanks for the comments! Updated w/ tests and docs.

alyssawilk reviewed (dismissed):
Yep, looks great.  Thanks for that extra test!

alyssawilk commented:
Unfortunately it looks like DCO bot froze.  Unfortunately I think the only way to unwedge that is do a main merge :-/

snowp commented on source/extensions/filters/http/fault/fault_filter.cc:
```diff
@@ -434,6 +434,15 @@ void FaultFilter::onDestroy() {
 }
 
+bool FaultFilter::activeFaults() {
+  // Check if there is a delay timer or response rate limiter still applying
```
nit: proper punctuation in comments

kailun-qin replied:
Fixed all, thanks.

snowp commented on source/extensions/filters/http/fault/fault_filter.cc:
```diff
@@ -434,6 +434,15 @@ void FaultFilter::onDestroy() {
 }
 
+bool FaultFilter::activeFaults() {
+  // Check if there is a delay timer or response rate limiter still applying
```
Does the response rate limiter go away here? Seems like it's coming from config so it should be constant? The comment reads a bit if it can't actually change

snowp commented on source/extensions/filters/http/fault/fault_filter.cc:
```diff
@@ -434,6 +434,15 @@ void FaultFilter::onDestroy() {
 
+bool FaultFilter::activeFaults() {
+  // Check if there is a delay timer or response rate limiter still applying
+  if (delay_timer_ != nullptr || fault_settings_->responseRateLimit() != nullptr) {
```
You can omit the if/else and just do `return delay_time_ != nullptr ...`

kailun-qin replied:
Updated, thanks.

snowp commented on source/extensions/filters/http/fault/fault_filter.cc:
```diff
@@ -445,6 +454,13 @@ void FaultFilter::postDelayInjection(const Http::RequestHeaderMap& request_heade
   } else {
+    // Should not continue to count as an active fault after the delay has elapsed
+    // if no other type of fault is active
+    if (!activeFaults() && fault_active_) {
```
Wouldn't `fault_active_` always be true here? And wouldn't the delay timer always be done at this point? Wondering if this could be simplified to just checking if there's a response fault configured

kailun-qin replied:
Yes, fault_active_ should always be true here and the delay timer should always be done at this point.
Personally I think it would be no harm to sanity check delay_timer_ and fault_active_ once again, and it might be easier to extend w/ a helper function. Though indeed, we have no other case at this moment so that a simplification of checks can be done.

snowp replied:
I think generally we'd use ASSERTs to verify expected program state instead of checking things we know to be true, just as a way to make sure that the program flows the way we expect it to. That way as the code changes we hit an ASSERT and can update the checks, ensuring that we update the logic to match the new changes.

kailun-qin replied:
Thanks for the suggestions. Updated w/ asserts.

snowp reviewed (changes requested):
Thanks for working on this! Some initial questions/comments

kailun-qin commented:
Updated with the simplification of checks and fixed some punctuation nits.
@alyssawilk @snowp Please kindly help review and let me know if you have any suggestion regarding the simplification.

kailun-qin commented:
> Updated with the simplification of checks and fixed some punctuation nits.
> @alyssawilk @snowp Please kindly help review and let me know if you have any suggestion regarding the simplification.

ping @snowp @alyssawilk, PTAL, thanks!

alyssawilk commented:
I already approved this one - @snowp have any more thoughts?

snowp reviewed (approved):
LGTM,thanks for the fix!
